docs/graphs/graph-cloning.md

```diff
@@ -0,0 +1,154 @@
+---
+id: graph-cloning 
+title: Graph-cloning 
+sidebar_label: Graph-cloning  
+description: "The program is to return a deep copy of the graph, preserving the structure and values of its nodes."  
+tags: [dsa, algorithms, graph]
+---
+
+### Problem Statement:
+You are given the reference of a node in a connected undirected graph. Return a deep copy (clone) of the graph. Each node in the graph contains a value and a list of its neighbors. There are n nodes in the graph each with a unique value from 0 to n-1.
+
+## Definition:
+### Input
+- A pointer/reference to a node of the original graph. Each node is represented as follows:
+  - An integer `val`, which is the unique value of the node.
+  - A list of pointers/references to neighboring nodes.
+
+### Output
+- A pointer/reference to the corresponding cloned node of the graph.
+
+### Constraints
+- The graph is guaranteed to be connected and undirected.
+- The node values are unique integers ranging from `0` to `n - 1`, where `n` is the number of nodes in the graph.
+
+### Approach
+To clone a graph, perform a depth-first search (DFS) or breadth-first search (BFS) starting from the given node, using a hash map to track visited nodes and ensure that each node is cloned only once, along with its neighbors.
+
+## Algorithm Steps
+
+1. **Check for Null Node**:
+   - If the input node is `null`, return `null`.
+
+2. **Check Visited Nodes**:
+   - If the node has already been cloned (exists in the visited map), return the clone from the map.
+
+3. **Create a Clone Node**:
+   - Create a new node with the same value as the original node.
+   - Store the clone in the visited map.
+
+4. **Clone Neighbors**:
+   - Iterate through each neighbor of the original node:
+     - Recursively call the clone function for each neighbor.
+     - Add the cloned neighbor to the `neighbors` list of the cloned node.
+
+5. **Return Cloned Node**:
+   - After cloning all neighbors, return the cloned node.
+
+## Time Complexity:
+- The time complexity of the graph cloning algorithm is `O(V + E)`, where `V` is the number of vertices and `E` is the number of edges. Each node and its neighbors are visited once during the DFS traversal, making the process linear with respect to the graph's size. The use of a map to track visited nodes ensures efficient lookups.
+
+### Sample Input:
+     //  0
+        / \
+       1---2
+        
```

Comment:
write inside of `<pre> ... </pre>` or  ```` ``` ... ``` ````

Reply:
Done